fladle-plugin/src/main/java/com/osacky/flank/gradle/FulladlePlugin.kt

```diff
@@ -166,3 +177,50 @@ fun writeAdditionalTestApps(strs: List<String>, flankGradleExtension: FlankGradl
     }
   )
 }
+
+val Project.isAndroidAppModule
+  get() = plugins.hasPlugin("com.android.application")
+val Project.isAndroidLibraryModule
+  get() = plugins.hasPlugin("com.android.library")
+
+// returns false if the module explicitly disabled testing or if it simply had no tests
+val Project.hasAndroidTest: Boolean
+  get() {
+    val fulladleModuleExtension = extensions.findByType(FulladleModuleExtension::class.java) ?: return false
+    if ((isAndroidLibraryModule || isAndroidAppModule) && !fulladleModuleExtension.enabled.get()) {
+      return false
+    }
+
+    val testedExtension = extensions.getByType<TestedExtension>()
+    var testsFound = true
+    testedExtension.testVariants.configureEach testVariant@{
+      if (!file("$projectDir/src/androidTest").exists()) {
+        println("Ignoring $name test variant in $path: No tests in $projectDir/src/androidTest")
+        testsFound = false
+      }
+      return@testVariant
+    }
+    return testsFound
+  }
+
+fun overrideRootLevelConfigs(flankGradleExtension: FlankGradleExtension, fulladleModuleExtension: FulladleModuleExtension) {
+  // if the root module overrode any value in its fulladleModuleConfig block
```

Comment:
this seems racy. why do we have to override these local configs from the root? what is the order of precedence?

Reply:
Since we have no way of specifying which module should end up as root, we can instead specify the `fulladleModuleConfig` in every module. Once a module ends up as a root-level module, if its values are overridden, they are used instead of root values.

Reply:
```suggestion
  // if the root module has any value set in its fulladleModuleConfig block
```

Reply:
Ok, I think I had a misunderstanding here in my orignal question. i think I understand better what his happening here.
The `FlankGradleExtension` is used at the root and the `FulladleModuleExtension` is per module.
Why do we always want to set the value for the `FulladleModuleExtension` with the value in the `FlankGradleExtension`?
For example, with `maxTestShards` we may have set a high value which is to be used for the main app apk but not the other modules. We would then have to override it in all the other modules?
The `debugApk` makes sense to me to override, but I think I don't fully understand the use case for `maxTestShards`. 

Also, wow would I opt out of overriding the value?

Reply:
Makes sense, it is only in the case where it is chosen as the root module that it overrides.